scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -250,25 +250,52 @@ def highly_variable_genes(
             'pass `inplace=False` if you want to return a `np.recarray`.')
 
     if batch_key is None:
-        df = _highly_variable_genes_single_batch(adata,
-                                                 min_disp=min_disp, max_disp=max_disp,
-                                                 min_mean=min_mean, max_mean=max_mean,
-                                                 n_top_genes=n_top_genes,
-                                                 n_bins=n_bins,
-                                                 flavor=flavor)
+        df = _highly_variable_genes_single_batch(
+            adata,
+            min_disp=min_disp, max_disp=max_disp,
+            min_mean=min_mean, max_mean=max_mean,
+            n_top_genes=n_top_genes,
+            n_bins=n_bins,
+            flavor=flavor,
+        )
     else:
         sanitize_anndata(adata)
         batches = adata.obs[batch_key].cat.categories
         df = []
         for batch in batches:
             adata_subset = adata[adata.obs[batch_key] == batch]
-            hvg = _highly_variable_genes_single_batch(adata_subset,
-                                                      min_disp=min_disp, max_disp=max_disp,
-                                                      min_mean=min_mean, max_mean=max_mean,
-                                                      n_top_genes=n_top_genes,
-                                                      n_bins=n_bins,
-                                                      flavor=flavor)
-            hvg['gene'] = adata.var_names.values
+
+            # Filter to genes that are in the dataset
+            genes_before = adata_subset.var_names
+            filter_genes(adata_subset, min_cells=1)
```

Comment:
This makes a copy of the whole `AnnData`, which isn't needed here. I'd follow @flying-sheep's recommendation to use `inplace=False` or calculate it directly. You can probably hide the logging output with a context manager if you want.

Reply:
There’s only one for filtering warnings, so I quickly built one:

https://github.com/theislab/scanpy/blob/e63c2df73fc324675b5cff52b444101852efe055/scanpy/_settings.py#L39-L46

Reply:
Thanks! Used that now.